drivers/gpio/gpio_axp192.c

```diff
@@ -0,0 +1,292 @@
+	struct i2c_dt_spec i2c;
+	const struct device *mfd;
+};
+
+struct gpio_axp192_data {
+	struct gpio_driver_data common;
+	sys_slist_t cb_list_gpio;
+};
+
+static int gpio_axp192_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	int ret;
+	uint8_t port_val;
+	const struct gpio_axp192_config *config = dev->config;
+
+	ret = mfd_axp192_gpio_read_port(config->mfd, &port_val);
+	if (ret == 0) {
+		*value = port_val;
+	}
+
+	return ret;
+}
+
+static int gpio_axp192_port_set_masked_raw(const struct device *dev, gpio_port_pins_t mask,
+					   gpio_port_value_t value)
+{
+	int ret;
+	const struct gpio_axp192_config *config = dev->config;
+
+	ret = mfd_axp192_gpio_write_port(config->mfd, value, mask);
+
+	return ret;
+}
+
+static int gpio_axp192_port_set_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_axp192_port_set_masked_raw(dev, pins, pins);
+}
+
+static int gpio_axp192_port_clear_bits_raw(const struct device *dev, gpio_port_pins_t pins)
+{
+	return gpio_axp192_port_set_masked_raw(dev, pins, 0);
+}
+
+static inline int gpio_axp192_configure(const struct device *dev, gpio_pin_t pin,
+					gpio_flags_t flags)
+{
+	const struct gpio_axp192_config *config = dev->config;
+	int ret;
+	enum axp192_gpio_func func;
+
+	if (pin >= AXP192_GPIO_NUM_PINS) {
+		LOG_ERR("Invalid gpio pin (%d)", pin);
+		return -EINVAL;
+	}
+
+	/* Configure pin */
+	LOG_DBG("Pin: %d / flags=0x%x", pin, flags);
+	if ((flags & GPIO_OUTPUT) != 0) {
+
+		/* Initialize output function */
+		func = AXP192_GPIO_FUNC_OUTPUT_LOW;
+		if ((flags & GPIO_OPEN_DRAIN) != 0) {
+			func = AXP192_GPIO_FUNC_OUTPUT_OD;
+		}
+		ret = mfd_axp192_gpio_func_ctrl(config->mfd, dev, pin, func);
+		if (ret != 0) {
+			return ret;
+		}
+
+		/* Set init value */
+		if ((flags & GPIO_OUTPUT_INIT_LOW) != 0) {
+			ret = mfd_axp192_gpio_write_pin(config->mfd, pin, false);
+		} else if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0) {
+			ret = mfd_axp192_gpio_write_pin(config->mfd, pin, true);
+		}
+	} else if ((flags & GPIO_INPUT) != 0) {
+
+		/* Initialize input function */
+		func = AXP192_GPIO_FUNC_INPUT;
+
+		ret = mfd_axp192_gpio_func_ctrl(config->mfd, dev, pin, func);
+		if (ret != 0) {
+			return ret;
+		}
+
+		/* Configure pull-down */
+		if ((flags & GPIO_PULL_UP) != 0) {
+			/* not supported */
+			LOG_ERR("Pull-Up not supported");
+			ret = -ENOTSUP;
+		} else if ((flags & GPIO_PULL_DOWN) != 0) {
+			/* out = 0 means pull-down*/
+			ret = mfd_axp192_gpio_pd_ctrl(config->mfd, pin, true);
+		} else {
+			ret = mfd_axp192_gpio_pd_ctrl(config->mfd, pin, false);
+		}
+	} else {
+		/* Neither input nor output mode is selected */
+		LOG_INF("No valid gpio mode selected");
+		ret = -ENOTSUP;
+	}
+
+	return ret;
+}
+
+static int gpio_axp192_port_toggle_bits(const struct device *dev, gpio_port_pins_t pins)
+{
+	int ret;
+	uint32_t value;
+
+	ret = gpio_axp192_port_get_raw(dev, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return gpio_axp192_port_set_masked_raw(dev, pins, ~value);
```

Comment:
Does this require locking?

Reply:
Very good catch. Your are absolutely right, there was missing protection for this rmw-action.
I also reviewed the mfd driver of possible issues, but couldn't find any further segments that also need protection.
Thank you for that valuable feedback.